otter/test/convergence/test_service.py

```diff
@@ -1,95 +1,425 @@
+        self.assertEqual(
+            self.successResultOf(d),
+            ('my-dispatcher',
+             Effect(CreateOrSet(path='/groups/divergent/tenant_group',
+                                content='dirty'))))
+        log.msg.assert_called_once_with(
+            'mark-dirty-success', tenant_id='tenant', group_id='group')
+
+    def test_error_marking_dirty(self):
+        """An error is logged when marking dirty fails."""
+        svc = ConvergenceStarter('my-dispatcher')
+        log = mock_log()
+
+        def perform(dispatcher, eff):
+            return fail(RuntimeError('oh no'))
+        d = svc.start_convergence(log, 'tenant', 'group', perform=perform)
+        self.assertEqual(self.successResultOf(d), None)
+        log.err.assert_called_once_with(
+            CheckFailureValue(RuntimeError('oh no')),
+            'mark-dirty-failure', tenant_id='tenant', group_id='group')
 
 
 class ConvergerTests(SynchronousTestCase):
+    """Tests for :obj:`Converger`."""
 
     def setUp(self):
-        clock = Clock()
-        self.kz_client = mock.Mock(Lock=LockMixin().mock_lock())
-        self.dispatcher = object()
-        self.converger = Converger(clock, self.kz_client, self.dispatcher)
-        self.state = GroupState('tenant-id', 'group-id', 'group-name',
-                                {}, {}, None, {}, False)
-        self.group = mock_group(self.state, 'tenant-id', 'group-id')
-        self.lc = {'args': {'server': {'name': 'foo'}, 'loadBalancers': []}}
+        self.dispatcher = _get_dispatcher()
+        self.log = mock_log()
+        self.buckets = range(10)
 
-    @mock.patch('time.time')
-    def test_converge(self, time):
+    def _pfactory(self, log, callable):
+        self.fake_partitioner = FakePartitioner(log, callable)
+        return self.fake_partitioner
+
+    def test_buckets_acquired(self):
         """
-        The ``converge`` method acquires a lock and performs the result of
-        :func:`execute_convergence` within that lock.
+        When buckets are allocated, the result of converge_all is performed.
         """
-        perform = mock.Mock()
+        def converge_all(log, group_locks, _my_buckets, all_buckets):
+            self.assertEqual(log, matches(IsBoundWith(system='converger')))
+            self.assertIs(group_locks, converger.group_locks)
+            self.assertEqual(_my_buckets, my_buckets)
+            self.assertEqual(all_buckets, self.buckets)
+            return Effect(Constant('foo'))
 
-        def execute_convergence(group, desired, lc, now, log):
-            return Effect(Constant((group, desired, lc, now, log)))
+        my_buckets = [0, 5]
+        converger = Converger(
+            self.log, self.dispatcher, self.buckets,
+            self._pfactory, converge_all=converge_all)
+
+        result = self.fake_partitioner.got_buckets(my_buckets)
+        self.assertEqual(self.successResultOf(result), 'foo')
```

Comment:
+1 for testing that all those assertions above actually get called - I often forget that :D